straxen/plugins/peaks/peak_subtyping.py

```diff
@@ -0,0 +1,621 @@
+            peaks[undefined_mask],
+            self.subtype_dtype,
+            self.large_s2_threshold, self.other_large_s2_fac,
+            self.s1_s2_window, self.after_s2_window_ext)
+
+        # check if there is any undefined peaks
+        n_undefined = (result['subtype'] == PeakSubtyping.Undefined).sum()
+        if n_undefined:
+            raise RuntimeError(f'We still find {n_undefined} peaks undefined!')
+
+        # sort the result by time
+        result = np.sort(result, order='time')
+
+        return result
+
+
+def limited_containers(anchor, length, limitation, forward=True):
+    """
+    Build container given anchor, length and limitation of end time.
+    The retuned containers should not be overlapping
+    :param anchor: sorted array of interval anchor points
+    :param length: length of each interval, len(anchor) == len(length)
+    """
+    assert np.all(length >= 0)
+    if not hasattr(length, '__len__'):
+        length = np.full(len(anchor), length)
+    # limitations must be sorted,
+    # and it is a combination of given limitation and anchor
+    # because we need to make sure the returned containers are not overlapping
+    # prevent numpy change the dtype to float
+    # limitation needs also be sorted
+    if len(limitation) > 0:
+        limitation = np.unique(np.hstack([anchor, limitation]))
+    else:
+        limitation = anchor.copy()
+    containers = np.zeros(len(anchor), dtype=strax.time_fields)
+    if forward:
+        containers['time'] = anchor
+        containers['endtime'] = _limited_containers_forward(anchor, length, limitation)
+    else:
+        containers['endtime'] = anchor
+        containers['time'] = _limited_containers_reverse(anchor, length, limitation)
+    assert np.all(containers['time'][1:] - containers['endtime'][:-1] >= 0)
+    return containers
+
+
+@numba.njit
+def _limited_containers_forward(time, length, limitation):
+    """
+    Construct the endtime in jit accelerated way
+    """
+    endtime = time + length
+    n = len(limitation)
+    i = 0
+    for j in range(len(time)):
+        while i < n - 1 and time[j] >= limitation[i]:
+            i += 1
+        endtime[j] = min(endtime[j], limitation[i])
+    return endtime
+
+
+@numba.njit
+def _limited_containers_reverse(endtime, length, limitation):
+    """
+    Construct the time in jit accelerated way
+    """
+    time = endtime - length
+    n = len(limitation)
+    i = n - 1
+    for j in np.arange(len(endtime))[::-1]:
+        while i > 0 and endtime[j] <= limitation[i]:
+            i -= 1
+        time[j] = max(time[j], limitation[i])
+    return time
+
+
+def combine_indices(result):
+    """
+    Combine the indices from touching_windows results
+    :param result: touching_windows results, each row is a pair of indices
+    """
+    if len(result) == 0:
+        return np.array([], dtype=np.int)
+    indices = []
+    for r in result:
+        indices.append(np.arange(r[0], r[1]))
```

Comment:
FYI: you could also concatenate instead of appending. Then you can remove the flattening operation in the next line.

Reply:
However, this might be less efficient. More efficient would be defining an array of the appropriate length first, adding information and then computing unqiue.